### ISSUE TITLE
copy_ami_snapshot fails for private/shared AMIs

### ISSUE THREAD

**Issue by infinitewarp** [cloudigrade/cloudigrade#357]
## Summary

The `copy_ami_snapshot` task fails at the `add_snapshot_ownership` call for `boto3` when the target AMI is shared with _but not owned by_ the target AWS account. This results in the snapshot failing to copy and the whole pipeline of async tasks failing.

## Steps to Reproduce

1. Register a cloud account that has a running instance based on a _private shared_ AMI that _is not_ owned by the target account. 
    - `ami-065d0be9883aa064a` is an example that belongs to our test environment's AWS account that has been _shared_ with the various individual developer "customer" AWS accounts.
1. Run Celery and watch for the processing of `copy_ami_snapshot`.

## Expected Result

- Task runs and completes normally.

## Actual Result

- Task fails and produces errors that look like:
```
[2018-06-18 23:03:04,358: INFO/MainProcess] Received task: account.tasks.copy_ami_snapshot[7a82f697-69d8-4ff0-ae6f-60e97b6bcf01]
[2018-06-18 23:03:04,421: INFO/ForkPoolWorker-9] Found credentials in environment variables.
[2018-06-18 23:03:04,443: INFO/ForkPoolWorker-9] Starting new HTTPS connection (1): sts.amazonaws.com
[2018-06-18 23:03:04,874: INFO/ForkPoolWorker-9] Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
[2018-06-18 23:03:05,142: INFO/ForkPoolWorker-9] Starting new HTTPS connection (1): ec2.us-east-1.amazonaws.com
[2018-06-18 23:03:05,369: INFO/ForkPoolWorker-9] Starting new HTTPS connection (1): sts.amazonaws.com
[2018-06-18 23:03:05,677: ERROR/ForkPoolWorker-9] Task account.tasks.copy_ami_snapshot[7a82f697-69d8-4ff0-ae6f-60e97b6bcf01] raised unexpected: RuntimeError("Unexpected AWS error <class 'botocore.exceptions.ClientError'>: An error occurred (InvalidSnapshot.NotFound) when calling the ModifySnapshotAttribute operation: None",)
Traceback (most recent call last):
  File "/Users/brasmith/projects/cloudigrade/cloudigrade/util/aws/helper.py", line 152, in wrapped
    result = original_function(*args, **kwargs)
  File "/Users/brasmith/projects/cloudigrade/cloudigrade/account/tasks.py", line 56, in copy_ami_snapshot
    aws.add_snapshot_ownership(snapshot)
  File "/Users/brasmith/projects/cloudigrade/cloudigrade/util/aws/ec2.py", line 169, in add_snapshot_ownership
    UserIds=[user_id]
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/boto3/resources/factory.py", line 520, in do_action
    response = action(self, *args, **kwargs)
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/boto3/resources/action.py", line 83, in __call__
    response = getattr(parent.meta.client, operation_name)(**params)
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidSnapshot.NotFound) when calling the ModifySnapshotAttribute operation: None

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/celery/app/trace.py", line 382, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/celery/app/trace.py", line 641, in __protected_call__
    return self.run(*args, **kwargs)
  File "/Users/brasmith/.virtualenvs/cloudigrade/lib/python3.6/site-packages/celery/app/base.py", line 467, in run
    return task._orig_run(*args, **kwargs)
  File "/Users/brasmith/projects/cloudigrade/cloudigrade/util/aws/helper.py", line 155, in wrapped
    raise RuntimeError(message)
RuntimeError: Unexpected AWS error <class 'botocore.exceptions.ClientError'>: An error occurred (InvalidSnapshot.NotFound) when calling the ModifySnapshotAttribute operation: None
```

## Additional context

@kholdaway and I found this when diagnosing local development issues. Giving the assumed role _all_ IAM permissions (including AdministratorAccess) didn't even help. So, it looks like we need another general approach for images that exist in this state. Maybe there's a different snapshot attribute we can set on the third-party snapshot? That seems unlikely but still worth checking. Maybe we need to copy the whole image to our account first before attempting to copy the snapshot? If this is the case, then we'll need to preserve some kind of mapping between the original AMI and the one we've copied so we can associate the copy's inspection results back to the original.

As cloudigrade developers, a workaround for this issue is to always run your customer account EC2 Instances backed by AMIs that belong to that account. Copy the shared images instead of launching directly from them.

**Comment by infinitewarp** [comment]
Well, it turns out we cannot copy the private shared AMI to our account directly for the same reason we cannot copy the snapshot. Copying an image requires access to be granted to the storage that backs it — either the EBS snapshot or the S3 bucket — which is the original problem all over again.

The only next solution I've seen is that we could first copy the snapshot _to the customer's account_, grant ourselves access to that, copy the snapshot to _our_ account, and delete the copy from the customer's account. This certainly is not ideal because it may incur additional costs to the customer for that intermediary copy.

**Comment by infinitewarp** [comment]
Dan and I talked through a few options for this problem, and I'm going to experiment with the aforementioned "make a copy in the customer's account" solution while he has a few discussions. Tomorrow afternoon maybe we'll have a new direction, but we have a path to attempt in the mean time.

**Comment by infinitewarp** [comment]
After realizing that adding `ec2:DeregisterImage` and `ec2:DeleteSnapshots` permissions could be a very scary and dangerous thing, I reached out to Dan for another brief discussion.

**tl;dr** is we may _not_ clean up after ourselves, but we need some way to bubble a notification up to the customer to indicate that we have created copies in their account that they should delete.

Conversation copied here:

> brad [1:34 PM]
> so, I just realized something as I was writing the cleanup part of this "make a copy in the customer's account" change
> In order to complete this operation, it means that when we give the customer the JSON blob to put in their AWS policy for us, we have to include in that set some permissions for `ec2:CopyImage`, `ec2:DeregisterImage`, and `ec2:DeleteSnapshot`.
> That's not a problem technically, but putting on my customer hat, I feel a panicked sweat come on when I think about giving someone else the ability to delete things in my account.
> It is necessary for us to be a "good neighbor" and clean up after ourselves, but I wonder if more customers will be turned off by that versus being told they need to clean up after us...?
> 
> danoz [1:38 PM]
> I agree fundamentally with reservations about delete.  I, myself, am nervous about what we might accidentally delete despite having no intention to do any such damage.
> 
> brad [1:39 PM]
> yes, collateral blast damage is a real risk too if we have a bug.
> 
> danoz [1:39 PM]
> I feel like the more straightforward solution here is to inform customers that if they're using shared images we have to pull a copy of that image in the account that we're monitoring.
> And alert them that image XYZ was copied to account ABC as image DEF
> 
> brad [1:41 PM]
> mmhmm mmhmm :thinking_face:
> 
> danoz [1:41 PM]
> That could get onerous - but
> I don't know enough yet about what we're going to find in customer accounts / situations.
> I may need to just escalate to Xav and say "I now need access to customer teams to proceed intelligently."
> 
> brad [1:42 PM]
> completely understood.
> 
> danoz [1:43 PM]
> For now, assume that we can let the copy persist and we'll alert.  Add a hook for "delete stuff here"
> That way if we find out that something like a button that says "I am about to delete image DEF (or a short list)" is OK, we have a place for it.
> 
> brad [1:44 PM]
> Okay, I was thinking something along those lines. I'll pull out the "clean up after yourself" bits for now and find a way to persist enough information to notify the customer that the copy exists.

**Comment by bclarhk** [comment]
Having a way for them to do that through Cloud Meter would be cleaner for customers that trust the app to do its job. Manually setting each account up could be tedious enough. Having to go back and clean up the mess we made - in every account - would not be a pleasant experience.

**Comment by kholdaway** [comment]
I think deleting images in a customer account is scary.  A bug could be destructive to their environment.

**Comment by ilyavw** [comment]
Perhaps make it conditional? Its more effort, but it allows the customer to choose. The safe approach: You'll have to clean stuff up in this edge case - Use this policy json. The "I trust you guys" approach: We'll tidy up for you - Use this policy json that has the scarier permissions.

**Comment by bclarhk** [comment]
> it allows the customer to choose.
👍

**Comment by tgerla** [comment]
Hey all, I've read through this and I have a few general thoughts. First, if we need to request additional scary permissions, I think that is OK. Perhaps we could tighten our permission requests down so that we only have permission to delete and deregister images and snapshots that are tagged a certain special way?

I think at a certain point we won't be able to avoid requesting powerful permissions from our customers, and they will have to trust us, and we'll have to trust ourselves, our test suites, and our code. As long as we are aware of the dangers going in, and focus on the safety of the routines exercising those permissions, I feel pretty comfortable doing it.

We can indeed present our customers with some alternatives if they are unable to grant us the various permissions. Some of the alternatives include:

1. Ask them to manually tag images that contain Red Hat software and we will trust them, and we will not need to crack open the image. The downside here is that we are again relying on self-reporting.

2. Require that our customer ask the upstream image owner to grant US permission. This sounds like a headache and is unlikely to work.

3. Request copy and snapshot permissions but not deregister/delete, causing crap to build up in the customer account. I think this is the least-appealing approach.

@dozdowski, I wonder if we could run these options and our concerns by a friendly pilot customer or two, and gauge their reaction?

**Comment by dozdowski** [comment]
The customer teams I've worked with so far haven't had technical knowledge in-depth enough to verify whether shared images are going to be a problem.  I've been escalating my way up to some folks who should be able to verify how much we'll need to deal with this, and can possibly weigh in on "If we need a temporary inspection copy we need to delete it - how does this make you feel?"

I'll add this to my pile of questions to ask customers when I start talking to them, rather than first-line account teams.

**Comment by infinitewarp** [comment]
The linked PRs add the functionality to make the copy into the customer's account since that's what (outside of this Github issue, at least) we discussed and deemed to be the best temporary solution. If or when we decide to abandon or modify that approach, it shouldn't be hard to rip, replace, or conditionalize what I've built here.

Technical demo: https://asciinema.org/a/189289

**Comment by ilyavw** [comment]
Looks good, verified.
